enhancements/baremetal-config.md

```diff
@@ -0,0 +1,153 @@
+---
+title: Baremetal IPI Config Enhancement
+authors:
+  - "@sadasu"
+reviewers:
+  - "@deads2k"
+  - "@hardys"
+  - "@enxebre"
+  - "@squeed"
+  - "@abhinavdahiya"
+
+approvers:
+  - "@deads2k"
+
+creation-date: 2019-10-23
+last-updated: 2019-10-23
+status: implemented
+see-also:
+  - "https://github.com/openshift/api/pull/480"  
+replaces:
+  - 
+superseded-by:
+  - 
+---
+
+# Config required for Baremetal IPI deployments
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Open Questions [optional]
+
+None.
+
+## Summary
+
+The configuration required to bring up a Baremetal IPI (metal3) cluster needs
+to be availabe in a Config Resource. The installer aleady adds some configuration
+to the Infrastructure Config CR. It is also aware of the additional config
+required during the deployment of a metal3 cluster. This enhancement proposes
+to extend the BareMetaPlatformStatus CR to hold these additional values needed
+for a metal3 deployment.
+
+## Motivation
+
+Adding these additional config items to the BareMetalPlatformStatus CR, allows
+versioning of these config items.
+
+### Goals
+
+The goal of this enhancement is to enable the installer to add vital config
+values for a Baremetal IPI deployment to a CR. The MAO will then be able to
+access these values when it kicks off a metal3 cluster.
+
+Update lifecycle for bootimage is an unsolved problem for all platforms. For
+more information, refer to : https://github.com/openshift/os/issues/381.
+
+### Non-Goals
+
+All config values except the RHCOSImageUrl are not expected to change after a
+metal3 cluster is up. Handling changes to this value after install time has
+not been explored and that is a non-goal at this time.
+
+### Proposal
+
+To support configuration of a Baremetal IPI (metal3) deployment, a set of new
+config items are required to be made availabe via a Config Resource. The
+BareMetalPlatformStatus is an ideal fit because it contains configuration
+that is specific to a Baremetal deployment.
+
+This proposal aims to add the following to BareMetalPlatformStatus:
+
+1. ProvisioningInterface : This is the interface name on the underlying
```

Comment:
Is this something like talk to this interface if you want to talk to the API, like `QEMU URL/amazonaws.com`

Also this seems like BareMetalHost Custom Resource should have this information, because i would assume users might want to have different interface on each baremetal-server that is used to talk to provisioning network??

Reply:
This is the physical nic used for the dedicated network on the nodes where the baremetal operator and Ironic runs. e.g eno1, eth3 or whatever.

This shouldn't be part of the BareMetalHost CRD because that describes the nodes to be provisioned, this is an implementation detail of how the provisioning service runs on the cluster.

Reply:
@abhinavdahiya  perhaps a bit more context would help here:

- We integrated the baremetal operator via the MAO in https://github.com/openshift/machine-api-operator/pull/302, and this starts some provisioning services in the same pod as the BMO (Ironic and dependencies).  To start these services we require some configuration data which is environment/hardware dependent but this data is *not* related to the actual hosts we deploy, it's required for the deployment services (to configure the PXE boot services primarily).

- Currently, we have to manually create a ConfigMap with that data ref https://github.com/openshift-metal3/dev-scripts/blob/master/metal3-config.yaml - this works but it's a barrier to users deploying via openshift-install without scripts/workarounds, essentially this is a gap from the previous MAO integration (and thus IMHO it's a bug not really an enhancement as such).

- We initially proposed to add the ConfigMap via the installer, since most of the data we need should be requested via the install-config at deploy time.  This was rejected in https://github.com/openshift/installer/pull/2149 by @wking with the suggestion this data should be added to the infrastructure config instead.

- Consequently we reworked the installer changes in https://github.com/openshift/installer/pull/2449 and added corresponding API changes in https://github.com/openshift/api/pull/480 and the MAO integration consumes this data in https://github.com/openshift/machine-api-operator/pull/402

AFAICS the changes proposed by @sadasu are aligned with the guidance previously provided by @wking, but perhaps you can provide more detailed guidance as to how we can refine this approach to make progress.

We've been trying to add this small amount of additional configuration data for nearly 3 months now, and it seems like conflicting feedback on the various PRs is part of the reason we're making such slow progress - perhaps you and @wking could chat and provide some very specific guidance on how to proceed?  Thanks!

Reply:
> Is this something like talk to this interface if you want to talk to the API, like `QEMU URL/amazonaws.com`
> 
> Also this seems like BareMetalHost Custom Resource should have this information, because i would assume users might want to have different interface on each baremetal-server that is used to talk to provisioning network??

@abhinavdahiya That is a good point. But, currently we are operating under the assumption that in a given installation, we would have baremetal servers that will all use an interface of the same name to be physically connected to the provisioning network.
Secondly, the BareMetalHost Custom Resourse currently contains information about baremetal hosts known to the BareMetal Operator. We also have an additional set of configs that are used by Ironic, which is a service that provisions baremetal servers and gets them ready for use as hosts and then as nodes. This config cannot be added to the BareMetalHost Custom Resource. 
So, instead of creating a brand new Custom Resourse, we are augmenting the BareMetalPlatformStatus Custom Resource. Agreed that adding this make the BareMetalPlatformStatus CR look a lot different from similar CRs used by other clouds.  But, in the Baremetal case, we are also responsible for provisioning of the underlying physical servers.

Reply:
An example value for this might be "eth1" or "ens8" or the name of a bridge (we use "provisioning" in the VM-based development environment for this work).

Reply:
So the provision network information is for the baremetal-controller so that it knows how to reach the provisioning network on the host it's running to do the provisioning of a new node?

Reply:
> So the provision network information is for the baremetal-controller so that it knows how to reach the provisioning network on the host it's running to do the provisioning of a new node?

Wait, why can't this just be a routing config on the controller node?  Controller: "I'm trying to reach the provisioning API at https://whatever.example.com", DNS: "whatever.example.com is 1.2.3.4", kernel: "take eth1 to reach 1.2.3.4".  Doesn't seem bare-metal-specific to me.

Reply:
A config similar to that might be the IngressIP and it is present in the BareMetalPlatformStatus CR. https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go#L176.

Reply:
> So the provision network information is for the baremetal-controller so that it knows how to reach the provisioning network on the host it's running to do the provisioning of a new node?

Not quite, it's so that the baremetal-operator pod can configure a dedicated interface on the node where that pod runs, this happens via https://github.com/openshift/ironic-static-ip-manager which is a bit hacky but works in the absence of an external DHCP server or any other external dependencies.

So to reply to @wking routing doesn't help us here, we have to bring up a nic on the node so we can serve DHCP/TFTP etc for pxe booting the nodes.

In future this may become optional, when we support use of external DHCP servers on the provisioning network, but right now it's required to make pxe boot work.

Reply:
> Controller: "I'm trying to reach the provisioning API at https://whatever.example.com"

FWIW I think this is a fundamental misunderstanding here - there is no existing provisioning API we have to reach - the metal3 pod deployed by the MAO runs that API, and most of the data under discussion here is just the configuration needed to make that hosted API work